### DISCUSSION TITLE
Conditional Searching

### DISCUSSION THREAD

**Discussion by ItsRobbAllen** [ravendb/ravendb#13995] [General]
Here's the scenario I am working with

A Unit is an abstract class from which we have a handful of concrete implementations (Project, Location, etc). Often, the client doesn't know information about the Unit, but they know something about the Company (which itself is an abstract class that also has several concrete implementations) that owns the Unit. 

I have indexes for both Units & Companies that uses `WhereEntityIs` to cover all types since they all share the same underlying fields.

When a client searches, they are saying _"Find me any unit that matches this search criteria _and_ find me all units that belong to any companies that match the search criteria."_

Every Unit has a reference to its `CompanyId` as there are too many units to store directly in the company record.

The code I have is as follows
```
var searchTerm = $"*Acme*";

var companyQuery = session.Query<Company>("Companies/SearchAllTypes")
    .Search(c => c.Name, searchTerm)
    .Search(c => c.Address.StreetAddress1, searchTerm)
    .Search(c => c.Address.StreetAddress2, searchTerm)
    .Search(c => c.Address.City, searchTerm);


var unitQuery = session.Query<Unit>("Units/SearchAllTypes")
    .Search(c => c.Name, searchTerm)
    .Search(c => c.Address.StreetAddress1, searchTerm)
    .Search(c => c.Address.StreetAddress2, searchTerm)
    .Search(c => c.Address.City, searchTerm)
    .Search(c => c.Number, searchTerm)
    .Search(u => u.CompanyId, companyQuery.Select(c => c.Id));
```

This works fine if the companyQuery returns any companies, but fails with `Cannot search on empty searchTerms array.` if there aren't any matches.

It really isn't a huge deal to execute the `companyQuery` & then modify the `unitQuery` if we get records back, but I'm trying to figure out if this can be done in a single trip to Raven.

**Comment by ayende** [comment]
Well, to start with, what is the logic of `.Search(u => u.CompanyId, companyQuery.Select(c => c.Id));` ?

You are doing full text search on the company id? Why no use `.In(u => u.CompanyId, companyQuery.Select(c => c.Id));`

That said, if you are searching by all of them in this manner, is there a reason not to have a single index covering both of them?
See: https://ayende.com/blog/152833/orders-search-in-ravendb


```
public class UnitsSearch : AbstractMultiMapIndexCreationTask
{
    public UnitsSearch()
    {
           AddMapForAll<UnitBase>(units => 
              from u in units select new { 
                  Query = new [] { 
                            u.Name, 
                            u.Address.StreetAddress1, 
                           u.Address.StreetAddress2, 
                           u.Address.City, 
                           u.Number, 
                           u.CompanyId
           });
    }

}
```

The above will search just on orders, but you can then apply that with `LoadDocument` to search on units by the related `Company` details.

**Reply by ItsRobbAllen** [reply]
Let's say there's 2 companies - _Rhino Tours_ `(companies/1-A)` and _Raven Construction_ `(companies/2-A)`

_Rhino Tours_ owns 25 units, _Raven_ has 3

One of the units, _Tasty-Treats_ `(units/123-A)`, owned by _Raven_ is located on _3417 Rhino Way_

A search for _rhino_ will give me `companies/1-A` matched on the company name and therefore I should get all 25 units that are part of `companies/1-A` AND `units/123-A` since the unit's address is also matched.

A search for _tasty_ would not match any companies but it would match `units/123-A` directly. In this case, using an `In` would return nothing since no company ids were found, and that's not what we need.

We actually find companies by several other criteria as well, such as contacts and various other bits of information, so there are several searches performed, but in the end they all just return a list of company ids so I simplified my code example above to not confuse matters.

Does that make sense?

**Reply by ayende** [reply]
The question is, what it is that you are trying to find. That is units, no?
Because it looks like you want to do the search term applied on both companies _and_ units at the same time, no?

In that case, can't you have an index like the one above, with the fields you want to search `UnitsQuery` and `CompanyQuery` and then your query will be:


`from index UnitsCompaniesSearch where search(UnitsQuery, $searchTerm) and search(CompanyQuery, $searchTerm)`

At which point you'll have matches only for those that have _both_ matches in units & company.

Using your example, you'll have a search for `Rhino` you'll get those that are in the company, but need something more to match the units, no?

Or you can use boosting on the `CompanyQuery` field to make it weight more. In general, that is likely to be more relevant if this is a search scenario. So you won't do filtering, but boosting based on how close that is to what you want.

A search on `Rhino` should first find units for `Rhino Tours` but also the `Rhino Way` unit (but last in the list) ?

**Comment by ItsRobbAllen** [comment]
> At which point you'll have matches only for those that have both matches in units & company.

That's the thing - I don't want units that have matches in both, I want units in _either_. 

"Show me all units that match the search criteria OR units that belong to companies that match the search criteria"

**Reply by ayende** [reply]
So, that would be:

`from index UnitsCompaniesSearch where search(UnitsQuery, $searchTerm) OR search(CompanyQuery, $searchTerm)`

No?

**Reply by ayende** [reply]
Note that I'm using `OR` here.
